File: src/alert_queries_runner.py
#!/bin/env python3

import json
import hashlib
import uuid
import datetime

from config import ALERTS_TABLE, RULES_SCHEMA, RESULTS_SCHEMA, ALERT_QUERY_POSTFIX
from helpers import log
from helpers.db import connect_and_execute, load_rules


GROUPING_PERIOD = 1 * 60 * 60  # Group events within one hour periods


# Grab alerts from past grouping_period amount of time
def get_recent_alerts(ctx, alert_type):
    alert_map = {}
    recent_alerts = ctx.cursor().execute(
        f"SELECT alert, counter FROM {ALERTS_TABLE} "
        f"WHERE try_cast(alert:EVENT_TIME::string as timestamp_ntz) >= DATEADD('second', (-1 * {GROUPING_PERIOD}), CURRENT_TIMESTAMP()) "
        f"  AND alert:AlertType = '{alert_type}';"
    ).fetchall()

    for alert in recent_alerts:
        current_alert = json.loads(alert[0])
        key = hashlib.md5((current_alert['OBJECT'] + current_alert['DESCRIPTION']).encode('utf-8')).hexdigest()
        alert_map[key] = [current_alert, alert[1], False]

    return alert_map


# Check if the proposed alert was already created recently, and update its counter
def alert_exists(alert_map, new_alert):
    uniq = new_alert['OBJECT'] + new_alert['DESCRIPTION']
    key = hashlib.md5(uniq.encode('utf-8')).hexdigest()
    if key in alert_map:
        alert_map[key][1] = alert_map[key][1] + 1
        alert_map[key][2] = True
        return True
    else:
        alert_map[key] = [new_alert, 1, False]
        return False


# After checking all recent alerts, update counter for alerts that were duplicated
def update_recent_alerts(ctx, alert_map):
    update_array = []
    update_array_length = 0
    for key in alert_map:
        if alert_map[key][2]:
            update_array.extend([alert_map[key][0]['ALERT_ID'], alert_map[key][1]])
            update_array_length = update_array_length + 1
    if update_array_length:
        format_string = ", ".join(["(%s, %s)"] * update_array_length)
        ctx.cursor().execute(f"CREATE TEMPORARY TABLE {RESULTS_SCHEMA}.counter_table(ALERT_ID string, COUNTER number);")
        ctx.cursor().execute(
            f"INSERT INTO {RESULTS_SCHEMA}.counter_table (ALERT_ID, COUNTER) VALUES {format_string};",
            update_array
        )
        ctx.cursor().execute(
            f"MERGE INTO {RESULTS_SCHEMA}.alerts s"
            f" USING {RESULTS_SCHEMA}.counter_table t"
            f" ON s.alert:ALERT_ID = t.ALERT_ID WHEN MATCHED THEN UPDATE"
            f" SET s.COUNTER = t.COUNTER;"
        )


def log_alerts(ctx, alerts):
    if len(alerts):
        print("Logging alerts...")
        format_string = ", ".join(["(%s)"] * len(alerts))
        try:
            ctx.cursor().execute((
                f'INSERT INTO {ALERTS_TABLE}(alert_time, alert) '
                f'SELECT PARSE_JSON(column1):ALERT_TIME, PARSE_JSON(column1) '
                f'FROM values {format_string};'),
                alerts)
        except Exception as e:
            log.fatal("Failed to log alert", e)
    else:
        print("No alerts to log.")


def log_failure(ctx, query_name, e):
    alerts = [json.dumps({
        'ALERT_ID': uuid.uuid4().hex,
        'QUERY_ID': '3a3d173a64ca4fcab2d13ac3e6d08522',
        'QUERY_NAME': 'Query Runner Failure',
        'ENVIRONMENT': 'Queries',
        'SOURCES': 'Query Runner',
        'ACTOR': 'Query Runner',
        'OBJECT': query_name,
        'ACTION': 'Query Execution',
        'TITLE': 'Query Runner Failure',
        'ALERT_TIME': str(datetime.datetime.utcnow()),
        'EVENT_TIME': str(datetime.datetime.utcnow()),
        'EVENT_DATA': f"The query '{query_name}' failed to execute with error: {e!r}",
        'DESCRIPTION': f"The query '{query_name}' failed to execute with error: {e!r}",
        'DETECTOR': 'Query Runner',
        'SEVERITY': 'High'
    })]
    try:
        log_alerts(ctx, alerts)
        log.fatal("Query failure successfully logged", e)
    except Exception as e:
        log.fatal("Failed to log query failure", e)


def snowalert_query(query_name: str):
    print(f"Received query {query_name}")

    ctx = connect_and_execute()

    try:
        query = f'''
            SELECT OBJECT_CONSTRUCT(*) from {RULES_SCHEMA}.{query_name}
            WHERE EVENT_TIME > dateadd(minute, -90, current_timestamp())
        '''
        results = ctx.cursor().execute(query).fetchall()
    except Exception as e:
        log_failure(ctx, query_name, e)

    print(f"Query {query_name} executed")
    return results, ctx


def process_results(results, ctx, query_name):
    alerts = []
    recent_alerts = get_recent_alerts(ctx, query_name)
    for res in results:
        jres = json.loads(res[0])
        if 'OBJECT' not in jres or 'DESCRIPTION' not in jres:
            log.error(f'OBJECT and DESCRIPTION required in {jres}')
            continue

        jres['ALERT_ID'] = uuid.uuid4().hex
        if not alert_exists(recent_alerts, jres):
            alerts.append(json.dumps(jres))
    log_alerts(ctx, alerts)
    update_recent_alerts(ctx, recent_alerts)


def query_for_alerts(query_name: str):
    results, ctx = snowalert_query(query_name)
    process_results(results, ctx, query_name)


def main():
    # Force warehouse resume so query runner doesn't have a bunch of queries waiting for warehouse resume
    ctx = connect_and_execute("ALTER SESSION SET USE_CACHED_RESULT=FALSE;")
    for query_name in load_rules(ctx, ALERT_QUERY_POSTFIX):
        query_for_alerts(query_name)


if __name__ == '__main__':
    main()

File: src/scripts/installer-queries/alert_queries_runner.js
// args

var FROM_TIME_SQL, TO_TIME_SQL, CUTOFF_MINUTES, QUERY_NAME;

CUTOFF_MINUTES = CUTOFF_MINUTES || 90;
FROM_TIME_SQL = FROM_TIME_SQL || `date_trunc(hours, current_timestamp) - interval '${CUTOFF_MINUTES} minutes'`;
TO_TIME_SQL = TO_TIME_SQL || 'current_timestamp';

// library

function exec(sqlText, binds=[]) {
  let retval = []
  const stmnt = snowflake.createStatement({sqlText, binds})
  const result = stmnt.execute()
  const columnCount = stmnt.getColumnCount();
  const columnNames = []
  for (let i = 1 ; i < columnCount + 1 ; i++) {
    columnNames.push(stmnt.getColumnName(i))
  }

  while(result.next()) {
    let o = {};
    for (let c of columnNames) {
      o[c] = result.getColumnValue(c)
    }
    retval.push(o)
  }
  return retval
}

function fillArray(value, len) {
  const arr = []
  for (var i = 0; i < len; i++) {
    arr.push(value)
  }
  return arr
}

const RUN_ID = Math.random().toString(36).substring(2).toUpperCase();
const NEW_ALERTS_TBL = `results.RUN_${RUN_ID}_${QUERY_NAME}`;

const CREATE_ALERTS_SQL = `CREATE TRANSIENT TABLE ${NEW_ALERTS_TBL} AS
SELECT OBJECT_CONSTRUCT(
         'ALERT_ID', UUID_STRING(),
         'QUERY_NAME', '${QUERY_NAME}',
         'QUERY_ID', IFNULL(QUERY_ID::VARIANT, PARSE_JSON('null')),
         'ENVIRONMENT', IFNULL(ENVIRONMENT::VARIANT, PARSE_JSON('null')),
         'SOURCES', IFNULL(SOURCES::VARIANT, PARSE_JSON('null')),
         'ACTOR', IFNULL(ACTOR::VARIANT, PARSE_JSON('null')),
         'OBJECT', IFNULL(OBJECT::VARIANT, PARSE_JSON('null')),
         'ACTION', IFNULL(ACTION::VARIANT, PARSE_JSON('null')),
         'TITLE', IFNULL(TITLE::VARIANT, PARSE_JSON('null')),
         'EVENT_TIME', IFNULL(EVENT_TIME::VARIANT, PARSE_JSON('null')),
         'ALERT_TIME', IFNULL(ALERT_TIME::VARIANT, PARSE_JSON('null')),
         'DESCRIPTION', IFNULL(DESCRIPTION::VARIANT, PARSE_JSON('null')),
         'DETECTOR', IFNULL(DETECTOR::VARIANT, PARSE_JSON('null')),
         'EVENT_DATA', IFNULL(EVENT_DATA::VARIANT, PARSE_JSON('null')),
         'SEVERITY', IFNULL(SEVERITY::VARIANT, PARSE_JSON('null')),
         'HANDLERS', IFNULL(OBJECT_CONSTRUCT(*):HANDLERS::VARIANT, PARSE_JSON('null'))
       ) AS alert
     , alert_time
     , event_time
     , 1 AS counter
FROM rules.${QUERY_NAME}
WHERE event_time BETWEEN ${FROM_TIME_SQL} AND ${TO_TIME_SQL}
;`;

const MERGE_ALERTS_SQL = `MERGE INTO results.alerts AS alerts USING (

  SELECT ANY_VALUE(alert) AS alert
       , SUM(counter) AS counter
       , MIN(alert_time) AS alert_time
       , MIN(event_time) AS event_time

  FROM ${NEW_ALERTS_TBL}
  GROUP BY alert:OBJECT, alert:DESCRIPTION

) AS new_alerts

ON (
  alerts.alert:OBJECT = new_alerts.alert:OBJECT
  AND alerts.alert:DESCRIPTION = new_alerts.alert:DESCRIPTION
  AND alerts.alert:EVENT_TIME > ${FROM_TIME_SQL}
)

WHEN MATCHED
THEN UPDATE SET counter = alerts.counter + new_alerts.counter

WHEN NOT MATCHED
THEN INSERT (alert, counter, alert_time, event_time)
  VALUES (
    new_alerts.alert,
    new_alerts.counter,
    new_alerts.alert_time,
    new_alerts.event_time
  )
;`;

return {
  'run_id': RUN_ID,
  'new_alerts_tbl': NEW_ALERTS_TBL,
  'create_alerts_result': exec(CREATE_ALERTS_SQL)[0],
  'merge_alerts_result': exec(MERGE_ALERTS_SQL)[0],
}
